### DISCUSSION TITLE
Type safety of Systems in Xstate?

### DISCUSSION THREAD

**Discussion by anovi** [statelyai/xstate#5063] [Q&A]
Hi there. I am interested in implementing the actor model in the front end of my application. But I am struggling with Systems in Xstate. In particular, how to use Systems with Typescript.

For example, I have a root machine with a setup that declares two actors:

```typescript
setup({
    actors:{
        actor1: childMachine1,
        actor2: childMachine2,
    }
})
```

At some of the states of the root machine, I invoke these child machines:
```typescript
…
invoke: [{
    src: 'actor1',
    systemId: 'actor1',
}],
…
```

Then, in another actor of the same system, I try to get this actor from the system:

```typescript
someState: {
    entry: ({ system }) => {
        const someActor1 = system.get('actor1');
            // do something with someActor1
        },
},
```

It works in runtime. But how do you get type safety with Systems? Because now it's `const someActor: any`. The documentation does not explain this. Am I missing something?

**Comment by with-heart** [comment]
Typed actor systems aren't currently supported. There are generic type parameters for a system in all the right places in the api but they're currently set to `AnyActorSystem` (`any`) rather than a specific actor system type.

I believe this is because there currently isn't a mechanism for defining a system as a whole. Right now actors have an implicit `system` which is just a system created for the root-level actor which is available to all descendent actors, but there's no way to define the shape of that system.

@davidkpiano critique my explanation?

**Reply by with-heart** [reply]
I think part of the complexity in implementing this is demonstrated in your example.

Your root machine defines actors. You invoke some of the actors some of the time. How should the types for the system reflect that? What if you define but never invoke any of the actors?
